components/io_expander/esp_io_expander/include/esp_io_expander.h

```diff
@@ -0,0 +1,225 @@
+     */
+    esp_err_t (*read_output_reg)(esp_io_expander_handle_t handle, uint32_t *value);
+
+    /**
+     * @brief Write value to direction register (mandatory)
+     *
+     * @note The value represents the diection of IO
+     * @note If there are multiple input registers in the device, their values should be spliced together in order to form the `value`.
+     *
+     * @param handle: IO Expander handle
+     * @param value: Register's value
+     *
+     * @return
+     *      - ESP_OK: Success, otherwise returns ESP_ERR_xxx
+     */
+    esp_err_t (*write_direction_reg)(esp_io_expander_handle_t handle, uint32_t value);
+
+    /**
+     * @brief Read value from directioin register (mandatory)
+     *
+     * @note The value represents the expected direction of IO
+     * @note This function can be implemented by reading the physical direction register, or simply by reading a variable that record the direction value (more faster)
+     * @note If there are multiple input registers in the device, their values should be spliced together in order to form the `value`.
+     *
+     * @param handle: IO Expander handle
+     * @param value: Register's value
+     *
+     * @return
+     *      - ESP_OK: Success, otherwise returns ESP_ERR_xxx
+     */
+    esp_err_t (*read_direction_reg)(esp_io_expander_handle_t handle, uint32_t *value);
+
+    /**
+     * @brief Reset the device to its initial state (mandatory)
+     *
+     * @note This function will reset all device's registers
+     *
+     * @param handle: IO Expander handle
+     *
+     * @return
+     *      - ESP_OK: Success, otherwise returns ESP_ERR_xxx
+     */
+    esp_err_t (*reset)(esp_io_expander_t *handle);
+
+    /**
+     * @brief Delete device (mandatory)
+     *
+     * @param handle: IO Expander handle
+     *
+     * @return
+     *      - ESP_OK: Success, otherwise returns ESP_ERR_xxx
+     */
+    esp_err_t (*del)(esp_io_expander_t *handle);
+
+    /**
+     * @brief Configuration structure
+     */
+    esp_io_expander_config_t config;
+};
+
+/**
+ * @brief Set the direction of a specific IO
+ *
+ * @param handle: IO Exapnder handle
+ * @param pin_num: IO num, valid from 0 to (`io_amount` - 1)
+ * @param is_output: true - output, false - input
+ *
+ * @return
+ *      - ESP_OK: Success, otherwise returns ESP_ERR_xxx
+ */
+esp_err_t esp_io_expander_set_dir(esp_io_expander_handle_t handle, uint8_t pin_num, bool is_output);
+
+/**
+ * @brief Get the direction of a specific IO
+ *
+ * @param handle: IO Exapnder handle
+ * @param pin_num: IO num, valid from 0 to (`io_amount` - 1)
+ * @param is_output: true - output, false - input
+ *
+ * @return
+ *      - ESP_OK: Success, otherwise returns ESP_ERR_xxx
+ */
+esp_err_t esp_io_expander_get_dir(esp_io_expander_handle_t handle, uint8_t pin_num, bool *is_output);
+
+/**
+ * @brief Set the output level of a specific IO
+ *
+ * @note Specific IO must be in input mode first, otherwise this function will return an error
```

Comment:
In input mode for set level?
Which error will be returned? Is there any specific error code for it?

Reply:
Oh, this is a careless mistake. Fixed and added explain for specific error code [here](https://github.com/espressif/esp-bsp/pull/86/commits/ca919496b317741e0e9ac3e8e858dbb289b5e385#diff-946a3784b32d4d423f23b0a03eb9d05a89128cc72555733cbd528e355395c4faL168-R201).